core/node/api_server/src/tx_sender/gas_estimation.rs

```diff
@@ -0,0 +1,511 @@
+        let vm_permit = sender.0.vm_concurrency_limiter.acquire().await;
+        let vm_permit = vm_permit.ok_or(SubmitTxError::ServerShuttingDown)?;
+
+        Ok(Self {
+            sender,
+            transaction,
+            state_override,
+            vm_permit,
+            fee_input,
+            base_fee,
+            gas_per_pubdata_byte,
+            max_gas_limit,
+            block_args,
+            protocol_version,
+        })
+    }
+
+    pub(super) fn adjust_transaction_fee(&mut self) {
+        match &mut self.transaction.common_data {
+            ExecuteTransactionCommon::L2(common_data) => {
+                common_data.fee.max_fee_per_gas = self.base_fee.into();
+                common_data.fee.max_priority_fee_per_gas = self.base_fee.into();
+            }
+            ExecuteTransactionCommon::L1(common_data) => {
+                common_data.max_fee_per_gas = self.base_fee.into();
+            }
+            ExecuteTransactionCommon::ProtocolUpgrade(common_data) => {
+                common_data.max_fee_per_gas = self.base_fee.into();
+            }
+        }
+    }
+
+    pub(super) async fn initialize(&self) -> Result<InitialGasEstimate, SubmitTxError> {
+        let operator_overhead = self.tx_overhead(self.max_gas_limit);
+
+        // When the pubdata cost grows very high, the total gas limit required may become very high as well. If
+        // we do binary search over any possible gas limit naively, we may end up with a very high number of iterations,
+        // which affects performance.
+        //
+        // To optimize for this case, we first calculate the amount of gas needed to cover for the pubdata. After that, we
+        // need to do a smaller binary search that is focused on computational gas limit only.
+        if self.transaction.is_l1() {
+            // For L1 transactions the pubdata priced in such a way that the maximal computational
+            // gas limit should be enough to cover for the pubdata as well, so no additional gas is provided there.
+            Ok(InitialGasEstimate {
+                total_gas_charged: None,
+                computational_gas_used: None,
+                operator_overhead,
+                gas_charged_for_pubdata: 0,
+            })
+        } else {
+            // For L2 transactions, we estimate the amount of gas needed to cover for the pubdata by creating a transaction with infinite gas limit,
+            // and getting how much pubdata it used.
+
+            let (result, _) = self
+                .unadjusted_step(self.max_gas_limit)
+                .await
+                .context("estimate_gas step failed")?;
+            // If the transaction has failed with such a large gas limit, we return an API error here right away,
+            // since the inferred gas bounds would be unreliable in this case.
+            // FIXME: double-check that returning early is OK
+            result.check_api_call_result()?;
```

Comment:
Subj. I *think* it's OK, at least if the transaction success / failure is monotonic w.r.t. the gas limit. In the old implementation, such a transaction would end binary search with the gas limit equal to the upper bound, and then would fail on `into_api_call_result()`. It's difficult to analyze implications in the general case though.

Reply:
I also think it's OK